include/dlaf/eigensolver/eigensolver.h

```diff
@@ -35,6 +35,15 @@ namespace dlaf::eigensolver {
 /// @param mat contains the Hermitian matrix A
 /// @param eigenvalues is a N x 1 matrix which on output contains the eigenvalues
 /// @param eigenvectors is a N x N matrix which on output contains the eigenvectors
```

Comment:
Some algorithms establish a precondition by talking about sizes in terms of `N` like here. Should we rather move this to a precondition and say `eigenvalues.rows() is equal to eigenvectors.rows()/cols()` or something like that? Or add the precondition _in addition_ to saying here that they're `N x 1` or `N x N`?

Reply:
I'd be in favor of moving this into a precondition. However, duplication of this information might lead to divergence.

Reply:
Not sure if doxigen likes this, but I would prefer to group the argument and its precondition to simplify the read. Moreover I would simplify the preconditions defining some variables. E.g. for the generalized eigensolver:
```
/// @param[in,out] mat_a contains the Hermitian matrix A, on output the value is destroyed
/// @pre mat_a is not distributed
/// @pre mat_a has size (N x N)
/// @pre mat_a has blocksize (NB x NB)
/// @pre mat_a has tilesize (NB x NB)
///
/// @param[in,out] mat_b contains the Hermitian positive definite matrix B on output the value is destroyed
/// @pre mat_b is not distributed
/// @pre mat_b has size (N x N)
/// @pre mat_b has blocksize (NB x NB)
/// @pre mat_b has tilesize (NB x NB)

/// @param[out] eigenvalues on output contains the eigenvalues
/// @pre eigenvalues is not distributed
/// @pre mat_b has size (N x 1)
/// @pre mat_b has blocksize (NB x 1)
/// @pre mat_b has tilesize (NB x 1)

/// @param[out] eigenvectors on output contains the eigenvectors
/// @pre eigenvectors is not distributed
/// @pre eigenvectors has size (N x N)
/// @pre eigenvectors has blocksize (NB x NB)
/// @pre eigenvectors has tilesize (NB x NB)
```
In this way it is much clearer what has to match among the different matrices.
What do you think?

Reply:
Exactly, giving "values" to the sizes is used to simplify writing the preconditions.
```
/// @pre mat_a has size (M x N)
...
/// @pre mat_b has size (N x M)
```
is immediately clear, while in the second case the sizes has to be reconstructed from the preconditions.
The first approach is more maintainable as if any of the precondition is relaxed it is easier to update the documentation.

Reply:
See https://github.com/eth-cscs/DLA-Future/pull/944/commits/54b6a7307d6502b52bd329553f1fe56d6ee7ed25 for potential new version. The grouping seems to work just fine:
![2023-08-30-110239_1669x1221_scrot](https://github.com/eth-cscs/DLA-Future/assets/42977/b1e99acd-2203-44a2-b9a2-2ac980928f32)
Slightly unsure what would be best to put on preconditions for the grids, especially the 1d grids. Let me know what you think would be best.

Reply:
For the grid I would use "mat is distributed according to grid" as two grids with the same size might not be compatible (e.g. row-major vs col-major grid).

Note: `eigenvalues` is not distributed, each rank has a full copy of it.

Reply:
Updated all of `eigensolver/eigensolver.h` in https://github.com/eth-cscs/DLA-Future/pull/944/commits/5622f8bd804a0b57a306ba448df019d8e1b2b962. @rasolca, @albestro does that look good? If yes, I think that'll generalize nicely to the other algorithms.

Reply:
My only concern is about the grid part of the precondition.
As said in my previous comment:
`/// @pre mat is distributed on an (NG x MG) grid` is incorrect and I would keep the previous statement `/// @pre mat is distributed according to grid`.

The reason is that the first statement would allow:
- reordered grids (e.g. col vs row -major grids)
- grids built on top of reordered communicators
- grids built on top of communicators with different ranks

Reply:
Sorry about that. I never staged that change... One more try: https://github.com/eth-cscs/DLA-Future/issues/966.